modules/xfeatures2d/src/daisy.cpp

```diff
@@ -0,0 +1,1754 @@
+            d0 = rx - hx; d1 = ry - hy;
+            radius = sqrt( d0*d0 + d1*d1 );
+            hradius[r] = quantize_radius( (float) radius );
+         }
+
+         ihx = (int)hx; if( hx - ihx > 0.5 ) ihx++;
+         ihy = (int)hy; if( hy - ihy > 0.5 ) ihy++;
+
+         if ( ! Point( ihx, ihy ).inside(
+                Rect( 0, 0, m_image.cols-1, m_image.rows-1 ) )
+            ) continue;
+
+         histogram = descriptor+region*m_hist_th_q_no;
+         ni_get_histogram( histogram, ihy, ihx, ishift, ptr + hradius[r]*m_cube_size );
+      }
+    }
+    return true;
+}
+
+inline void DAISY_Impl::initialize_single_descriptor_mode( )
+{
+    initialize();
+    compute_smoothed_gradient_layers();
+}
+
+inline void DAISY_Impl::set_parameters( )
+{
+    m_grid_point_number = m_rad_q_no * m_th_q_no + 1; // +1 is for center pixel
+    m_descriptor_size = m_grid_point_number * m_hist_th_q_no;
+
+    for( int i=0; i<360; i++ )
+    {
+      m_orientation_shift_table[i] = i/360.0 * m_hist_th_q_no;
+    }
+    m_layer_size = m_image.rows*m_image.cols;
+    m_cube_size = m_layer_size*m_hist_th_q_no;
+
+    compute_cube_sigmas();
+    compute_grid_points();
+}
+
+// set/convert image array for daisy internal routines
+// daisy internals use CV_32F image with norm to 1.0f
+inline void DAISY_Impl::set_image( InputArray _image )
+{
+    // release previous image
+    // and previous workspace
+    reset();
+    // fetch new image
+    Mat image = _image.getMat();
+    // image cannot be empty
+    CV_Assert( ! image.empty() );
+    // clone image for conversion
+    if ( image.depth() != CV_32F ) {
+
+      m_image = image.clone();
+      // convert to gray inplace
+      if( m_image.channels() > 1 )
+          cvtColor( m_image, m_image, COLOR_BGR2GRAY );
+      // convert and normalize
+      m_image.convertTo( m_image, CV_32F );
+      m_image /= 255.0f;
+    } else
+      // use original user supplied CV_32F image
+      // should be a normalized one (cannot check)
+      m_image = image;
+}
+
+
+// -------------------------------------------------
+/* DAISY interface implementation */
+
+// keypoint scope
+void DAISY_Impl::compute( InputArray _image, std::vector<KeyPoint>& keypoints, OutputArray _descriptors )
+{
+    // do nothing if no image
+    if( _image.getMat().empty() )
+      return;
+
+    set_image( _image );
+
+    // whole image
+    m_roi = Rect( 0, 0, m_image.cols, m_image.rows );
+
+    // get homography
+    Mat H = m_h_matrix;
+
+    // convert to double if case
+    if ( H.depth() != CV_64F )
+        H.convertTo( H, CV_64F );
+
+    set_parameters();
+
+    initialize_single_descriptor_mode();
+
+    // allocate array
+    _descriptors.create( (int) keypoints.size(), m_descriptor_size, CV_32F );
+
+    // prepare descriptors
+    Mat descriptors = _descriptors.getMat();
+    descriptors.setTo( Scalar(0) );
+
+    // iterate over keypoints
+    // and fill computed descriptors
+    if ( H.empty() )
```

Comment:
looks like there is almost complete code duplication, whereas the only difference (if I understood correctly) is point_transform_via_homography() call. If you pass NULL pointer as "double* H" and check it before calling point_transform_via_homography() - that will be very efficient, because the condition will be predicted perfectly by CPU and you loose may be just a few CPU clocks per descriptor (and computing a single DAISY descriptor takes 100's of clocks)